tailfs/compositefs/compositefs_test.go

```diff
@@ -0,0 +1,507 @@
+
+	// Make some files, use perms 0666 as lowest common denominator that works
+	// on both UNIX and Windows.
+	err := os.WriteFile(filepath.Join(dir1, "file1.txt"), []byte("12345"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.WriteFile(filepath.Join(dir2, "file2.txt"), []byte("54321"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// make some directories
+	err = os.Mkdir(filepath.Join(dir1, "dir1"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = os.Mkdir(filepath.Join(dir2, "dir2"), 0666)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if opts == nil {
+		opts = &Opts{}
+	}
+	if opts.Logf == nil {
+		opts.Logf = t.Logf
+	}
+	clock := tstest.NewClock(tstest.ClockOpts{Start: time.Now()})
+	opts.Clock = clock
+
+	fs := New(opts)
+	fs.AddChild(&Child{Name: "remote4", FS: &closeableFS{webdav.Dir(dir2)}})
+	fs.SetChildren(&Child{Name: "remote2", FS: webdav.Dir(dir2)},
+		&Child{Name: "remote3", FS: &closeableFS{webdav.Dir(dir2)}},
+	)
+	fs.AddChild(&Child{Name: "remote1", FS: webdav.Dir(dir1)})
+	fs.RemoveChild("remote3")
+
+	child, ok := fs.GetChild("remote1")
+	if !ok || child == nil {
+		t.Fatal("unable to GetChild(remote1)")
+	}
+	child, ok = fs.GetChild("remote2")
+	if !ok || child == nil {
+		t.Fatal("unable to GetChild(remote2)")
+	}
+	child, ok = fs.GetChild("remote3")
+	if ok || child != nil {
+		t.Fatal("should have been able to GetChild(remote3)")
+	}
+	child, ok = fs.GetChild("remote4")
+	if ok || child != nil {
+		t.Fatal("should have been able to GetChild(remote4)")
+	}
+
+	return fs, dir1, dir2, clock, func() {
+		defer l1.Close()
+		defer os.RemoveAll(dir1)
+		defer l2.Close()
+		defer os.RemoveAll(dir2)
+	}
+}
+
+func stat(t *testing.T, path string) fs.FileInfo {
+	fi, err := os.Stat(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return fi
+}
+
+func startRemote(t *testing.T) (net.Listener, string) {
+	dir, err := os.MkdirTemp("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	l, err := net.Listen("tcp", "127.0.0.1:")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	h := &webdav.Handler{
+		FileSystem: webdav.Dir(dir),
+		LockSystem: webdav.NewMemLS(),
+	}
+
+	s := &http.Server{Handler: h}
+	go s.Serve(l)
+
+	return l, dir
+}
+
+func infosEqual(t *testing.T, expected, actual fs.FileInfo) {
+	if expected.Name() != actual.Name() {
+		t.Errorf("expected name: %v   got: %v", expected.Name(), actual.Name())
+	}
+	if expected.Size() != actual.Size() {
+		t.Errorf("expected Size: %v   got: %v", expected.Size(), actual.Size())
+	}
+	if !expected.ModTime().Truncate(time.Second).UTC().Equal(actual.ModTime().Truncate(time.Second).UTC()) {
+		t.Errorf("expected ModTime: %v   got: %v", expected.ModTime(), actual.ModTime())
+	}
+	if expected.IsDir() != actual.IsDir() {
+		t.Errorf("expected IsDir: %v   got: %v", expected.IsDir(), actual.IsDir())
+	}
+}
+
+// closeableFS is a webdav.FileSystem that implements io.Closer()
+type closeableFS struct {
+	webdav.FileSystem
+}
+
+func (cfs *closeableFS) Close() error {
+	log.Println("I'm closing!")
```

Comment:
Leftover?

If we want to log it, we could also bundle the `*testing.T` in:

```go
type closeableFS struct {
  t *testing.T
  webdav.FileSystem
}

func (f *closeableFS) Close() error {
  f.t.Log("Closeable filesystem closing")
  return nil
}
```

but YMMV.

Reply:
Fixed